### PR TITLE
Validation and row editing

### LINKED ISSUE #27: 3.1 Custom validators
All default data types should be supported. Default data types are:

- strings
- numbers
- dates

Will use the same API as the "Ability to make some fields mandatory" to flag invalid cells. 

Demo example should be provided

### PR DESCRIPTION
**Pull request by Toshegg**
Validation and row editing along with implementation of validations for basic renderers

Fixes #27.

### CODE REVIEW

**Comment by Toshegg** [comment]
Needs some work to make column hide when the grid is not editable.

**Comment by Toshegg** [comment]
Fixed tests, added editable attribute

File: px-data-grid-cell-content-wrapper.html [hunk 1]
```diff
@@ -53,6 +53,17 @@
 
           this.addEventListener('mouseover', this._fireCellHoverEvent);
           this.addEventListener('mouseout', this._fireCellUnhoverEvent);
+          document.addEventListener('editing-item-changed', (event) => {
```
Comment (alump) [inline px-data-grid-cell-content-wrapper.html]: Should this be called editted-item-changed (also maybe property should be called editted-item)

File: px-data-grid-date-renderer.html [hunk 2]
```diff
@@ -33,7 +55,14 @@
         }
 
         _getFormattedDate(value) {
-          return (new Date(value)).toString();
+          return this.value.length > 0 ? (new Date(value)).toString() : '';
```
Comment (alump) [inline px-data-grid-date-renderer.html]: Can value be undefined?

Comment (alump) [reply px-data-grid-date-renderer.html]: Also new Date(value) will throw exception if not valid? Should this have try catch?

File: px-data-grid-number-renderer.html [hunk 3]
```diff
@@ -1,17 +1,37 @@
...
-    <style>
+    <style include="px-data-grid-styles">
       :host {
         display: flex;
         align-items: center;
+        flex-grow: 1;
+      }
+
+      input[invalid] {
+        outline: 1px solid red;
```
Comment (alump) [inline px-data-grid-number-renderer.html]: There is already class name for input in invalid state? Check their website.

File: px-data-grid-number-renderer.html [hunk 4]
```diff
@@ -31,6 +51,13 @@
         static get is() {
           return 'px-data-grid-number-renderer';
         }
+
+        validate() {
+          return {
+            valid: !this.column.required || this.value.toString().length > 0,
+            message: 'Value can not be empty'
```
Comment (alump) [inline px-data-grid-number-renderer.html]: use translate

File: px-data-grid-string-renderer.html [hunk 5]
```diff
@@ -1,17 +1,38 @@
...
-    <style>
+    <style include="px-data-grid-styles">
       :host {
         display: flex;
         align-items: center;
+        flex-grow: 1;
+      }
+
+      input[invalid] {
+        outline: 1px solid red;
```
Comment (alump) [inline px-data-grid-string-renderer.html]: Use classname from predix theme

File: px-data-grid-string-renderer.html [hunk 6]
```diff
@@ -31,6 +52,13 @@
         static get is() {
           return 'px-data-grid-string-renderer';
         }
+
+        validate() {
+          return {
+            valid: !this.column.required || this.value.length > 0,
+            message: 'Value can not be empty'
```
Comment (alump) [inline px-data-grid-string-renderer.html]: Use translate

File: px-data-grid.html [hunk 7]
```diff
@@ -1282,6 +1344,44 @@
             this._vaadinGrid.detailsOpenedItems = [];
           }
         }
+
+        _editingItemObserver(editingItem, oldV) {
+          if (typeof oldV === 'undefined') {
+            return;
+          }
+
+          this.dispatchEvent(new CustomEvent('editing-item-changed', {
```
Comment (alump) [inline px-data-grid.html]: editted-item

File: px-data-grid.html [hunk 8]
```diff
@@ -607,6 +640,23 @@
               type: String,
               value: null
             },
+
+            /**
+             * Stores item, that is currently in edit mode.
+             */
+            _editingItem: {
```
Comment (alump) [inline px-data-grid.html]: edittedItem?

File: test/helpers.html [hunk 9]
```diff
@@ -62,8 +62,8 @@
     return grid.shadowRoot.querySelector('[slot="' + slotName + '"]');
   }
 
-  function getCellContent(grid, cell) {
-    cell = getCell(grid, cell);
+  function getCellContent(grid, _cell) {
```
Comment (alump) [inline test/helpers.html]: _ predix in parameter name?

Comment (Toshegg) [reply test/helpers.html]: As you can see, we are setting `cell` variable inside the function itself, so we should not have the same names for function's parameter and internal variable.

Comment (alump) [reply test/helpers.html]: Ok. _ prefix is ok.

File: px-data-grid-date-renderer.html (hunk 2 above)
Comment (Toshegg) [reply px-data-grid-date-renderer.html]: Done.

File: px-data-grid-number-renderer.html (hunk 3 above)
Comment (Toshegg) [reply px-data-grid-number-renderer.html]: Done.

File: px-data-grid-number-renderer.html (hunk 4 above)
Comment (Toshegg) [reply px-data-grid-number-renderer.html]: Done.

File: px-data-grid-string-renderer.html (hunk 5 above)
Comment (Toshegg) [reply px-data-grid-string-renderer.html]: Done.

File: px-data-grid-date-renderer.html [hunk 10]
```diff
@@ -1,17 +1,35 @@
...
     <div id="value">[[_getFormattedDate(value)]]</div>
     <div id="editingTemplate">
-      <input type="date" value="{{value::change}}">
+      <input id="editingTemplateInput" class$="{{getClasses(validationResult)}}" type="date" value="{{value::change}}">
+      <px-tooltip
+        for="editingTemplateInput"
+        orientation="top"
+        tooltip-message="[[validationResult.message]]"
+        opened="[[!validationResult.valid]]"
+        ignore-target-events="true">
```
Comment (alump) [inline px-data-grid-date-renderer.html]: minor thing, but is polymer way to do this just to define ignore-target-events attribute without value

File: px-data-grid-number-renderer.html [hunk 11]
```diff
@@ -1,17 +1,34 @@
...
     <div id="value">[[value]]</div>
     <div id="editingTemplate">
-      <input type="number" value="{{value::change}}">
+      <input id="editingTemplateInput" class$="{{getClasses(validationResult)}}" type="number" value="{{value::change}}">
+      <px-tooltip
+        for="editingTemplateInput"
+        orientation="top"
+        tooltip-message="[[validationResult.message]]"
+        opened="[[!validationResult.valid]]"
+        ignore-target-events="true">
```
Comment (alump) [inline px-data-grid-number-renderer.html]: just attribute, no value

File: px-data-grid-string-renderer.html [hunk 12]
```diff
@@ -1,17 +1,34 @@
...
     <div id="value">[[value]]</div>
     <div id="editingTemplate">
-      <input type="text" value="{{value::change}}">
+      <input id="editingTemplateInput" class$="{{getClasses(validationResult)}}" type="text" value="{{value::change}}">
+      <px-tooltip
+        for="editingTemplateInput"
+        orientation="top"
+        tooltip-message="[[validationResult.message]]"
+        opened="[[!validationResult.valid]]"
+        ignore-target-events="true">
```
Comment (alump) [inline px-data-grid-string-renderer.html]: just attribute, no value

**Comment by limonte** [comment]
Rebase to master

**Pull request merged**